### PR TITLE
[Outlining] LLVM Suffix Tree

### PR DESCRIPTION
**Pull request by ashleynh**
This PR adds LLVM's suffix tree data structure to Binaryen. This suffix tree is implemented using Ukkonen's algorithm for linear-time suffix tree construction, and is intended for fast substring queries.

Note: All of the .h and .cpp files included are from LLVM. These files were copied directly instead of imported into our existing LLVM integration (in third_party/) to avoid bumping the commit hash and avoid the potential for complications with upstream changes.

### CODE REVIEW

File: src/support/CMakeLists.txt [hunk 1]
```diff
@@ -14,4 +14,15 @@ set(support_SOURCES
...
+#[[ The below condition is intended for removal once the suffix_tree and suffix_tree_node source files no longer depend on LLVM code in the third_party folder ]]
+if(EMSCRIPTEN)
 add_library(support OBJECT ${support_SOURCES})
+else()
+set(support_with_suffix_tree_SOURCES
+  suffix_tree.cpp
+  suffix_tree_node.cpp
+  ${support_SOURCES}
+)
+add_library(support OBJECT ${support_with_suffix_tree_SOURCES})
```
Comment (tlively) [inline src/support/CMakeLists.txt:27]: I _think_ that we usually have two spaces of indent for code inside the if-else.

File: src/support/CMakeLists.txt [hunk 2]
```diff
@@ -14,4 +14,15 @@ set(support_SOURCES
   utilities.cpp
   ${support_HEADERS}
 )
+
+#[[ The below condition is intended for removal once the suffix_tree and suffix_tree_node source files no longer depend on LLVM code in the third_party folder ]]
```
Comment (tlively) [inline src/support/CMakeLists.txt:18]: Wrap to 80 chars, please.

**Review by tlively — APPROVED** [review]
LGTM % making the CMake beautiful, which it turns out I have opinions about after all.

File: src/support/CMakeLists.txt (hunk 2 above)
Comment (ashleynh) [reply src/support/CMakeLists.txt]: done

File: src/support/CMakeLists.txt (hunk 1 above)
Comment (ashleynh) [reply src/support/CMakeLists.txt]: done

**Pull request merged**